File: ECommerceApp/SubCategoriesTableViewController.m
//
//  SubCategoriesTableViewController.m
//  ECommerceApp
//
//  Created by Prateek Raj on 20/07/18.
//  Copyright © 2018 Prateek Raj. All rights reserved.
//

#import "SubCategoriesTableViewController.h"
#import "DetailViewController.h"

@interface SubCategoriesTableViewController () {
    
    SubCategoriesTableViewController *subCategoriesTVC;
    
    DetailViewController *detailsVC;
    
    Categories *mainCategory;
    
    NSArray *productsArray;
    NSArray *categoriesArray;
}

@end

@implementation SubCategoriesTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Uncomment the following line to preserve selection between presentations.
    // self.clearsSelectionOnViewWillAppear = NO;
    
    // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
    // self.navigationItem.rightBarButtonItem = self.editButtonItem;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

-(void)loadCategoryDetails:(Categories *)category{
    
    mainCategory = category;
    
    [self.tableView reloadData];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    
    categoriesArray = mainCategory.childCategories.allObjects;
    
    if (categoriesArray.count > 0)
        return categoriesArray.count;
    else
        return 1;
}

-(CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    
    return 44;
}

- (nullable UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section {
    
    UIView *view = [[UIView alloc]initWithFrame:CGRectMake(0, 0, tableView.frame.size.width, 44)];
    
    view.tag = section;
    [view setUserInteractionEnabled:YES];
    
    UILabel *label = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, view.frame.size.width, 40)];
    [label setFont:[UIFont boldSystemFontOfSize:17]];
    label.textAlignment = NSTextAlignmentCenter;
    
     [view addSubview:label];
    
    if (mainCategory.numberofChildCategories > 0) {
        
        NSArray *childCategories = mainCategory.childCategories.allObjects;
        
        ChildCategories *childCategory = childCategories[section];
        
        Categories *category = [DATAMANAGER getArrayforEntity:@"Categories"
                                          filterwithPredicate:[NSString stringWithFormat:@"categoryId == %d",childCategory.categoryId]
                                                  sortWithKey:nil isAscending:NO].lastObject;
        
        if (category.numberofChildCategories > 0) {
            
            UITapGestureRecognizer *tapGesture = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(didSelectSectionHeder:)];
            
            [view addGestureRecognizer:tapGesture];
            [view setBackgroundColor:[UIColor cyanColor]];
            
            UIImageView *arrow = [[UIImageView alloc] initWithFrame:CGRectMake(tableView.frame.size.width-60, 10, 33, 33)];
            
            arrow.image = [UIImage imageNamed:@"rightArrow"];
            
            [view addSubview:arrow];
        } else {
            [view setBackgroundColor:[UIColor lightGrayColor]];
        }
        
        label.text = category.name;
    } else {
        
        [view setBackgroundColor:[UIColor lightGrayColor]];
        label.text = mainCategory.name;
    }
    return view;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    
    if (mainCategory.numberofChildCategories > 0) {
        
        NSArray *childCategories = mainCategory.childCategories.allObjects;
        
        ChildCategories *childCategory = childCategories[section];
        
        Categories *category = [DATAMANAGER getArrayforEntity:@"Categories"
                                          filterwithPredicate:[NSString stringWithFormat:@"categoryId == %d",childCategory.categoryId]
                                                  sortWithKey:nil isAscending:NO].lastObject;
        
        return category.numberOfProducts;
    } else {
        
        productsArray = mainCategory.products.allObjects;
        return productsArray.count;
        
    }
}

-(UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"CATEGORYDETAILCELL"];
    
    if (!cell)
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"CATEGORYDETAILCELL"];
    
    if (mainCategory.numberofChildCategories > 0) {
        
        NSArray *childCategories = mainCategory.childCategories.allObjects;
        
        ChildCategories *childCategory = childCategories[indexPath.section];
        
        Categories *categories = [DATAMANAGER getArrayforEntity:@"Categories"
                                            filterwithPredicate:[NSString stringWithFormat:@"categoryId == %d",childCategory.categoryId]
                                                    sortWithKey:nil isAscending:NO].lastObject;
        
        Product *product = categories.products.allObjects[indexPath.row];
        
        cell.textLabel.text = product.name;
    } else {
        
        NSArray *productsArray = mainCategory.products.allObjects;
        Product *product = productsArray[indexPath.row];
        
        cell.textLabel.text = product.name;
    }
    
    return cell;
}

- (void)didSelectSectionHeder:(UITapGestureRecognizer *)tapGesture {
    
    if (!subCategoriesTVC)
        subCategoriesTVC = [APPDELEGATE.storyBoard instantiateViewControllerWithIdentifier:@"SubCategoriesTableViewController"];
    
    NSArray *subviews = tapGesture.view.subviews;
    
    UILabel *textLabel = nil;
    
    for (UIView *subview in subviews) {
        
        if ([subview isKindOfClass:[UILabel class]]) {
            
            textLabel = (UILabel *)subview;
            break;
        }
    }
    
    NSString *categoryName = textLabel.text;
    
    Categories *category = [DATAMANAGER getArrayforEntity:@"Categories"
                                      filterwithPredicate:[NSString stringWithFormat:@"name MATCHES '%@'",categoryName]
                                              sortWithKey:nil isAscending:NO].lastObject;
    
    [subCategoriesTVC loadCategoryDetails:category];
    [self.navigationController pushViewController:subCategoriesTVC animated:YES];
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    
    if (!detailsVC)
        detailsVC = [APPDELEGATE.storyBoard instantiateViewControllerWithIdentifier:@"DetailViewController"];
    
    UITableViewCell *cell = [tableView cellForRowAtIndexPath:indexPath];
    
    Product *product = [DATAMANAGER getArrayforEntity:@"Product"
                                      filterwithPredicate:[NSString stringWithFormat:@"name MATCHES '%@'",cell.textLabel.text]
                                              sortWithKey:nil isAscending:NO].lastObject;
    
    [detailsVC loadProductDetails:product];
    
    [self.navigationController pushViewController:detailsVC animated:YES];
    
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}


@end

File: ECommerceApp/SubCategoriesTableViewController.swift
//
//  SubCategoriesTableViewController.swift
//  ECommerceApp
//
//  Created by Prateek Raj on 28/07/18.
//  Copyright © 2018 Prateek Raj. All rights reserved.
//

import UIKit

class SubCategoriesTableViewController: UITableViewController {
    
    var subCategoriesTVC: SubCategoriesTableViewController!
    var detailsVC: DetailViewController!
    var mainCategory: Categories!
    var productsArray: [Product] = []
    var categoriesArray: [Categories] = []
    var selectedIndex: IndexPath!
    var selectedSection: Int!
    
    func loadCategoryDetails(_ category: Categories) {
        mainCategory = category
        tableView.reloadData()
    }
    
    func getProductsForSection(_ section: Int) -> [Product] {
        var category: Categories?
        if mainCategory.numberofChildCategories > 0 {
            category = mainCategory.childrens?.allObjects[section] as? Categories
        } else {
            category = mainCategory
        }
        return category?.products?.allObjects as? [Product] ?? []
    }
    
    @objc func didSelectSectionHeder(_ tapGesture: UITapGestureRecognizer) {
        if let tag = tapGesture.view?.tag {
            selectedSection = tag
            performSegue(withIdentifier: "SubCategoriesVC", sender: self)
        }
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if let identifier = segue.identifier {
            if identifier == "showDetailVC" {
                let detailVC = segue.destination as! DetailViewController
                var product: Product? = nil
                let products = getProductsForSection(selectedIndex.section)
                if products.count > selectedIndex.row {
                    product = products[selectedIndex.row]
                }
                
                if let product = product {
                    detailVC.loadProductDetails(product)
                }
                tableView.deselectRow(at: selectedIndex, animated: false)
            } else if identifier == "SubCategoriesVC" {
                guard mainCategory.numberofChildCategories > selectedSection, let category = mainCategory.childrens?.allObjects[selectedSection] as? Categories else { return }
                let subCategoriesTVC = segue.destination as! SubCategoriesTableViewController
                subCategoriesTVC.loadCategoryDetails(category)
            }
        }
    }
}

typealias TableViewMethodss = SubCategoriesTableViewController

extension TableViewMethodss {
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        return mainCategory.numberofChildCategories > 0 ? Int(mainCategory.numberofChildCategories) : 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return getProductsForSection(section).count
    }
    
    override func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 44;
    }
    
    override func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        let view = UIView(frame: CGRect(x: 0, y: 0, width: tableView.frame.size.width, height: 44))
        view.tag = section
        view.isUserInteractionEnabled = true
        
        let label = UILabel(frame: CGRect(x: 0, y: 0, width: view.frame.size.width, height: 40))
        label.font = UIFont.boldSystemFont(ofSize: 17)
        label.textAlignment = .center
        view.addSubview(label)
        
        if (mainCategory.childCategories?.count ?? 0) > 0 {
            let category = mainCategory.childrens?.allObjects[section] as? Categories
            if (category?.childCategories?.count ?? 0) > 0 {
                view.backgroundColor = .cyan
                let tapGesture = UITapGestureRecognizer(target: self, action: #selector(SubCategoriesTableViewController.didSelectSectionHeder(_:)))
                
                let arrow = UIImageView(frame: CGRect(x: tableView.frame.size.width - 60, y: 10, width: 33, height: 33))
                arrow.image = UIImage(named: "rightArrow")
                view.addSubview(arrow)
                view .addGestureRecognizer(tapGesture)
            } else {
                view.backgroundColor = .lightGray
            }
            label.text = category?.name
        } else {
            view.backgroundColor = .lightGray
            label.text = mainCategory.name
        }
        return view
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "CATEGORYDETAILCELL") ?? UITableViewCell(style: .default, reuseIdentifier: "CATEGORYDETAILCELL")
        var product: Product?
        if mainCategory.numberofChildCategories > 0 {
            let childCategories = mainCategory.childrens?.allObjects
            let category = childCategories?[indexPath.section] as? Categories
            let products = category?.products?.allObjects
            product = products?[indexPath.row] as? Product
        } else {
            let products = mainCategory.products?.allObjects
            product = products?[indexPath.row] as? Product
        }
        cell.textLabel?.text = product?.name
        return cell
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        selectedIndex = indexPath
        performSegue(withIdentifier: "showDetailVC", sender: self)
    }
    
}
